lib/rouge/lexers/livescript.rb

```diff
@@ -0,0 +1,308 @@
+        rule(%r(^(?=\s|/))) { push :slash_starts_regex }
+        mixin :comments
+        mixin :whitespace
+
+        # list of words
+        rule %r/(<\[)(.*?)(\]>)/m do
+          groups Punctuation, Str, Punctuation
+        end
+
+        # function declarations
+        rule %r/!\s*function\b/, Keyword::Declaration
+        rule %r/!?[-~]>|<[-~]!?/, Keyword::Declaration
+
+        # switch arrow
+        rule %r/(=>)/, Keyword
+
+        # prototype attributes
+        rule %r/(::)(#{id})/ do
+          groups Punctuation, Name::Attribute
+          push :id
+        end
+        rule %r/(::)(#{int})/ do
+          groups Punctuation, Num::Integer
+          push :id
+        end
+
+        # instance attributes
+        rule %r/(@)(#{id})/ do
+          groups Name::Variable::Instance, Name::Attribute
+          push :id
+        end
+        rule %r/([.])(#{id})/ do
+          groups Punctuation, Name::Attribute
+          push :id
+        end
+        rule %r/([.])(\d+)/ do
+          groups Punctuation, Num::Integer
+          push :id
+        end
+        rule %r/#{id}(?=\s*:[^:=])/, Name::Attribute
+        rule %r/(&)(#{id})?/ do
+          groups Name::Builtin, Name::Attribute
+        end
+
+        # operators
+        rule %r(
+          [+][+]|--|&&|\b(and|x?or|is(nt)?|not)\b(?!-[a-zA-Z]|_)|[|][|]|
+          [.]([|&^]|<<|>>>?)[.]|\\(?=\n)|[.:]=|<<<<?|<[|]|[|]>|
+          (<<|>>|==?|!=?|[-<>+*%^/~?])=?
+        )x, Operator, :slash_starts_regex
+
+        # switch case
+        rule %r/[|]|\bcase(?=\s)/, Keyword, :switch_underscore
+
+        rule %r/@/, Name::Variable::Instance
+        rule %r/[.]{3}/, Punctuation
+        rule %r/:/, Punctuation
+
+        # keywords
+        rule %r/#{id}/ do |m|
+          if self.class.loop_control_keywords.include? m[0]
+            token Keyword
+            push :loop_control
+            next
+          elsif self.class.keywords.include? m[0]
+            token Keyword
+          elsif self.class.constants.include? m[0]
+            token Name::Constant
+          elsif self.class.builtins.include? m[0]
+            token Name::Builtin
+          elsif self.class.declarations.include? m[0]
+            token Keyword::Declaration
+          elsif /^[A-Z]/.match(m[0]) && /[^-][a-z]/.match(m[0])
+            token Name::Class
+          else
+            token Name::Variable
+          end
+          push :id
+        end
+
+        # punctuation and brackets
+        rule %r/\](?=[!?.]|#{id})/, Punctuation, :id
+        rule %r/[{(\[;,]/, Punctuation, :slash_starts_regex
+        rule %r/[})\].]/, Punctuation
+
+        # literals
+        rule %r/#{int_number}[.]#{int}/, Num::Float
+        rule %r/0x[0-9A-Fa-f]+/, Num::Hex
+        rule %r/#{int}/, Num::Integer
+
+        # strings
+        rule %r/"""/ do
+          token Str
+          push do
+            rule %r/"""/, Str, :pop!
+            rule %r/"/, Str
+            mixin :double_strings
+          end
```

Comment:
This is an interesting approach. I don't know if I recall having seen the creation of an 'anonymous' (for want of a better term) state via a block like this. I'm a little curious as to any performance hit (e.g. does this involve the creation of a new block, and all the objects associated with it, on every lex?). I think it's probably fine as-is and something we can revisit in the future if people start noticing issues.

Reply:
To be clear, you don't need to change this. I wanted to leave the comment for future reference.